cli/tools/test/reporters/junit.rs

```diff
@@ -167,44 +187,238 @@ impl TestReporter for JunitTestReporter {
     &mut self,
     elapsed: &Duration,
     tests: &IndexMap<usize, TestDescription>,
-    _test_steps: &IndexMap<usize, TestStepDescription>,
+    test_steps: &IndexMap<usize, TestStepDescription>,
   ) -> anyhow::Result<()> {
     let mut suites: IndexMap<String, quick_junit::TestSuite> = IndexMap::new();
     for (id, case) in &self.cases {
-      if let Some(test) = tests.get(id) {
-        suites
-          .entry(test.location.file_name.clone())
-          .and_modify(|s| {
-            s.add_test_case(case.clone());
-          })
-          .or_insert_with(|| {
-            quick_junit::TestSuite::new(test.location.file_name.clone())
-              .add_test_case(case.clone())
-              .to_owned()
-          });
-      }
+      let abs_filename = match (tests.get(id), test_steps.get(id)) {
+        (Some(test), _) => &test.location.file_name,
+        (_, Some(step)) => &step.location.file_name,
+        (None, None) => {
+          unreachable!("Unknown test ID '{id}' provided");
+        }
+      };
+
+      let filename = to_relative_path_or_remote_url(&self.cwd, abs_filename);
+
+      suites
+        .entry(filename.clone())
+        .and_modify(|s| {
+          s.add_test_case(case.clone());
+        })
+        .or_insert_with(|| {
+          let mut suite = quick_junit::TestSuite::new(filename);
+          suite.add_test_case(case.clone());
+          suite
+        });
     }
 
     let mut report = quick_junit::Report::new("deno test");
-    report.set_time(*elapsed).add_test_suites(
-      suites
-        .values()
-        .cloned()
-        .collect::<Vec<quick_junit::TestSuite>>(),
-    );
+    report
+      .set_time(*elapsed)
+      .add_test_suites(suites.into_values());
 
-    if self.path == "-" {
+    if self.output_path == "-" {
       report
         .serialize(std::io::stdout())
         .with_context(|| "Failed to write JUnit report to stdout")?;
     } else {
-      let file = crate::util::fs::create_file(&PathBuf::from(&self.path))
-        .context("Failed to open JUnit report file.")?;
+      let file =
+        crate::util::fs::create_file(&PathBuf::from(&self.output_path))
+          .context("Failed to open JUnit report file.")?;
       report.serialize(file).with_context(|| {
-        format!("Failed to write JUnit report to {}", self.path)
+        format!("Failed to write JUnit report to {}", self.output_path)
       })?;
     }
 
     Ok(())
   }
 }
+
+#[derive(Debug, Default)]
+struct TestNameTree(IndexMap<usize, TestNameTreeNode>);
+
+impl TestNameTree {
+  fn new() -> Self {
+    Self(IndexMap::new())
```

Comment:
Maybe pre-allocate to something like 256 to avoid excessive reallocations?

Reply:
Good point! fixed in [14d89d2](https://github.com/denoland/deno/pull/22797/commits/14d89d281015b73b99c7cb0304b4b1b04e8ee112)